### PR TITLE
[HLSL] Add ByteAddressBuffer, RWByteAddressBuffer and RasterizerOrderedByteAddressBuffer definitions to HLSLExternalSemaSource #113477

### LINKED ISSUE #58654: [HLSL] support ByteAdressBuffer
Create the ByteAddressBuffer, RWByteAddressBuffer, and RasterizerOrderedByteAddressBuffer templates in HLSLExternalSemaSource, which each contain a handle that will be lowered to a dx.RawBuffer target extension type.

### PR DESCRIPTION
**Pull request by joaosaffran**
This is the first one in a series of PRs adding the requirements for #58654 

This PR adds `ByteAddressBuffer`, `RWByteAddressBuffer ` and `RasterizerOrderedByteAddressBuffer ` definitions as well as their handle lowering to `dx.RawBuffer`.

closes #58654

### CODE REVIEW

File: clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-library -x hlsl -ast-dump -DEMPTY %s | FileCheck -check-prefix=EMPTY %s
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-library -x hlsl -ast-dump %s | FileCheck %s
+
```
Comment (hekota) [inline clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl:3]: FYI, we are planning to refactor the resource buffer AST tests, see here: https://github.com/llvm/llvm-project/pull/113643#discussion_r1833370629

Maybe there could be a single AST test with multiple run lines that would tests all of the raw buffer combinations using defines and different check-prefixes. Do you want to update this test to be ready for adding the next raw buffer variants or wait until the next buffer class is added?

File: clang/test/CodeGenHLSL/builtins/ByteAddressBuffer-contructor.hlsl [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -emit-llvm -disable-llvm-passes -o - %s | FileCheck %s --check-prefixes=CHECK,CHECK-DXIL
```
Comment (hekota) [inline clang/test/CodeGenHLSL/builtins/ByteAddressBuffer-contructor.hlsl:1]: Rename to ByteAddressBuffer**s**-constructor**s**.hlsl - test cases for other byte address buffers will be added here later on.

File: clang/lib/AST/DeclBase.cpp [hunk 3]
```diff
@@ -1474,6 +1474,11 @@ DeclContext *DeclContext::getPrimaryContext() {
   case Decl::ObjCCategoryImpl:
     return this;
 
+  case Decl::CXXRecord:
+    if (auto *OPD = dyn_cast<CXXRecordDecl>(this))
+      if (auto *Def = OPD->getDefinition())
+        return Def;
+    return this;
   default:
```
Comment (hekota) [inline clang/lib/AST/DeclBase.cpp:1482]: Why is this necessary?

Comment (joaosaffran) [reply clang/lib/AST/DeclBase.cpp:1482]: This is required to fix for these tests to pass:
```
  Clang :: AST/HLSL/ByteAddressBuffer-AST.hlsl
  Clang :: AST/HLSL/OutArgExpr.hlsl
  Clang :: AST/HLSL/pch.hlsl
  Clang :: AST/HLSL/pch_hlsl_buffer.hlsl
  Clang :: AST/HLSL/pch_with_buf.hlsl
  Clang :: CodeGenHLSL/builtins/ByteAddressBuffers-constructors.hlsl
```

Those test fail with the error message below if such change is not present:
```
llvm/include/llvm/Support/Casting.h:662: decltype(auto) llvm::dyn_cast(From *) [To = clang::TagType, From = const clang::Type]: Assertion `detail::isPresent(Val) && "dyn_cast on a non-existent value"' failed.
```

File: clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl (hunk 1 above)
Comment (joaosaffran) [reply clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl:3]: I refactor the test a little to make it more generic. However, I feel more comfortable making this change once we have more buffer classes, this way I will have more code to create patterns of.

Comment (hekota) [reply clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl:3]: I think it would be better if you commit the test in the original form and update it to cover more buffer types once you have more code the see the patterns. I have some ideas how to do it, let's chat later.

File: clang/lib/AST/DeclBase.cpp (hunk 3 above)
Comment (joaosaffran) [reply clang/lib/AST/DeclBase.cpp:1482]: I am not 100% sure, but my understanding is, `ByteAddressBuffer` is the first buffer to not use `addSimpleTemplateParams`. Such method, calls `Builder.addTypeParameter`, which modifies the declaration. 

@bogner  @llvm-beanz, do any of you can correct/confirm this fix, please?

Comment (hekota) [reply clang/lib/AST/DeclBase.cpp:1482]: Ok, that explains why the *pch* tests are now failing - they include the new buffer types.

I believe the default case in `DeclContext::getPrimaryContext` should be covering the CXXRecordDecl case as well. Since it hits the assert, my guess is that the CXXRecordDecl is probably not set up as expected. Maybe the definition is not complete? What is the difference between the records that use `addSimpleTemplateParams` and this one that does not when executing `DeclContext::getPrimaryContext`?

Comment (joaosaffran) [reply clang/lib/AST/DeclBase.cpp:1482]: The issue I was facing in the default case of `DeclContext::getPrimaryContext` was when calling `Tag->getTypeForDecl()` (line 1491). For `ByteAdrressBuffer` such method was returning `nullptr`.

For example, when calling `getTypeForDecl` for `StructuredBuffer` I get the response below:
```
InjectedClassNameType 0x5555668d3e20 'StructuredBuffer<element_type>' dependent
`-CXXRecord 0x5555668d3a60 'StructuredBuffer'
```

In contrast, if I call the same method for `ByteAddressBuffer` I get:
```
<<<NULL>>>
```

File: clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// There should be no more occurrences of ByteAddressBuffer
+// EMPTY-NOT: {{[^[:alnum:]]}}ByteAddressBuffer
+
+#ifndef EMPTY
+
+ByteAddressBuffer Buffer;
+
+#endif
+
+// CHECK: CXXRecordDecl 0x{{[0-9A-Fa-f]+}} <<invalid sloc>> <invalid sloc> implicit referenced <undeserialized declarations> class ByteAddressBuffer definition
```
Comment (bob80905) [inline clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl:17]: You should rebase / update the branch, and you'll find that this test fails.
Update class ByteAddressBuffer to typename ByteAddressBuffer, and it should succeed.

File: clang/lib/AST/DeclBase.cpp (hunk 3 above)
Comment (hekota) [reply clang/lib/AST/DeclBase.cpp]: Right, because the `TypeForDecl` is not set for `ByteAddressBuffer`. We need to make sure it gets set, otherwise the decl is not valid and it is going to bite us again someplace else.

We are creating the record in `BuiltinTypeDeclBuilder` constructor with `DelayTypeCreation` flag. For template classes the type gets created in `finalizeTemplateArgs` in the `ASTContext::getInjectedClassNameType` call, but that does not happed for non-template classes. We need to make sure it gets created and assigned, for example by adding `ASTContext::getTypeDeclType` call to the `BuiltinTypeDeclBuilder` destructor:

```
  ~BuiltinTypeDeclBuilder() {
    if (!Record->getTypeForDecl())
      Record->getASTContext().getTypeDeclType(Record,
                                              Record->getPreviousDecl());
    ... rest of the destructor code ....
  }

```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 5]
```diff
@@ -67,6 +67,7 @@ struct BuiltinTypeDeclBuilder {
     Record = CXXRecordDecl::Create(AST, TagDecl::TagKind::Class, HLSLNamespace,
                                    SourceLocation(), SourceLocation(), &II,
                                    PrevDecl, true);
+    Record->startDefinition();
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:70]: We should not be starting the definition here - this constructor just creates the declaration.

File: clang/test/CodeGenHLSL/builtins/ByteAddressBuffers-constructors.hlsl [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -emit-llvm -disable-llvm-passes -o - %s | FileCheck %s --check-prefixes=CHECK,CHECK-DXIL
+// RUN-DISABLED: %clang_cc1 -triple spirv-vulkan-library -x hlsl -emit-llvm -disable-llvm-passes -o - %s | FileCheck %s --check-prefixes=CHECK,CHECK-SPIRV
+
+// NOTE: SPIRV codegen for resource types is not yet implemented
+
+ByteAddressBuffer Buffer: register(t0);
+
+// CHECK: "class.hlsl::ByteAddressBuffer" = type { target("dx.RawBuffer", i8, 0, 0) }
```
Comment (hekota) [inline clang/test/CodeGenHLSL/builtins/ByteAddressBuffers-constructors.hlsl:8]: Add test cases for RWByteAddressBuffer and RasterizerOrderedByteAddressBuffer here as well.

File: clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl [hunk 7]
```diff
@@ -0,0 +1,49 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-library -x hlsl -ast-dump -DEMPTY \
```
Comment (hekota) [inline clang/test/AST/HLSL/ByteAddressBuffer-AST.hlsl:1]: Since the test now covers multiple buffers please rename it to ByteAddressBuffer**s**-AST.hlsl and update the PR title and description.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 8]
```diff
@@ -91,6 +94,18 @@ struct BuiltinTypeDeclBuilder {
       HLSLNamespace->addDecl(Record);
   }
 
+  CXXRecordDecl *finalizeForwardDeclaration() {
+    // Force the QualType to be generated for the record declaration. In most
```
Comment (inbelic) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:98]: Which of the added testcases would fail without this? Otherwise, could we add a testcase that would fail without this.

Comment (joaosaffran) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:98]: There are some existing tests that failed and motivated this fix:
```
  Clang :: AST/HLSL/ByteAddressBuffer-AST.hlsl
  Clang :: AST/HLSL/OutArgExpr.hlsl
  Clang :: AST/HLSL/pch.hlsl
  Clang :: AST/HLSL/pch_hlsl_buffer.hlsl
  Clang :: AST/HLSL/pch_with_buf.hlsl
  Clang :: CodeGenHLSL/builtins/ByteAddressBuffers-constructors.hlsl
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 9]
```diff
@@ -540,10 +555,10 @@ struct BuiltinTypeMethodBuilder {
 
     // create method decl
     auto *TSInfo = AST.getTrivialTypeSourceInfo(MethodTy, SourceLocation());
-    Method =
-        CXXMethodDecl::Create(AST, DeclBuilder.Record, SourceLocation(),
-                              NameInfo, MethodTy, TSInfo, SC_None, false, false,
-                              ConstexprSpecKind::Unspecified, SourceLocation());
+    Method = CXXMethodDecl::Create(
+        AST, DeclBuilder.finalizeForwardDeclaration(), SourceLocation(),
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:559]: This change is not necessary - methods are added when the type is getting completed on its first use, which is long after the forward declaration was created.

**Review by hekota — APPROVED** [review]
LGTM!
